### ISSUE TITLE
Add "local" XRFrameOfReference

### ISSUE THREAD

**Issue by Artyom17** [immersive-web/webxr#389]
@casseveritt recently brought a good issue up. With all the positional tracking untethered devices, like HoloLens or Lenovo Solo you may need to render UI elements in such way that the user doesn't end up inside of them. Even if you initially render something on the distance from the current user's location, the user still may come "into" the UI element. Of course, you always may say that the experience could be responsible to recalculate the UI elements' positions in the world coordinate system; however, if we rely on each apps' (experience) logic then the behavior will be inconsistent between them.
So the idea is, we don't want apps to have to figure out where the user wants UI-related stuff. We want to have a system query where apps can ask for the current "local origin". And when the user does a long-press of a specific button on a controller to recenter or whatever, what it does is update that local origin.
Since this is a system-level setting, all apps would be working with the same location / orientation. So overlays like UI or panels in the VR space, would draw around where the user has said his / her center is. But that wouldn't change the coordinates that the tracking system sends you by default. So stuff you want to draw around the user, you do by drawing it around the "local" pose. But apps that want to draw stuff in the physical world will need to know about both local space and the room space. By providing a query for local space, we don't require apps to try to come up with their own idea of where the user wants local stuff drawn. And, in particular, we don't have to worry about every app (experience) coming up with different ways to communicate this.

So, the idea is to add 'local' XRFrameOfReference type, where the player's position will be (0,0,0) at the time of pose reset, for example, when user press and hold a specific button on the controller.

Any thoughts on that?

**Comment by speigg** [comment]
Why not just use the eye-level frame, since the UA should be able to let the user reset that frame?

**Comment by casseveritt** [comment]
Is the eye level frame always intended to be centered around the user?  If "eye-level" is synonymous with the local frame, that could work.  (Though the "local" name I think captures what is intended more clearly.)

**Comment by speigg** [comment]
I agree, I think “local” is a better name for the “eye-level” frame—that’s what we used to call it in our Argon project, anyways. But yes, the “eye-level” frame should generally be “near” the user, while the “head-model” frame follows the user around.

**Comment by casseveritt** [comment]
Ok, I think the name "local" will be a better match for OpenXR, if that matters.

Stage frames don't have an implicit correspondence to the local frame though.  Rather apps have to explicitly ask for their relation (the transform between them).  Stage space's origin is always at the center of the stage, with the edges axis aligned.

**Comment by speigg** [comment]
@cassevereitt, you’re right, perhaps only the emulated stage would have an implicit correspondence.

**Comment by speigg** [comment]
@thetuvix I like your idea, but assuming that both the “local” and “stage” frame are available, isn’t the “local-floor” frame simply derivable from the “local” and “stage” frames? in other words, can’t an app just project the “local” frame onto the horizontal plane described by the “stage” frame in order to find the “local-floor”?
